app/src/main/java/at/bitfire/icsdroid/ui/screen/EditCalendarScreen.kt

```diff
@@ -0,0 +1,220 @@
+import androidx.compose.material3.IconButton
+import androidx.compose.material3.Scaffold
+import androidx.compose.material3.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.platform.LocalContext
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import androidx.lifecycle.viewmodel.compose.viewModel
+import at.bitfire.icsdroid.R
+import at.bitfire.icsdroid.db.entity.Subscription
+import at.bitfire.icsdroid.model.CredentialsModel
+import at.bitfire.icsdroid.model.EditCalendarModel
+import at.bitfire.icsdroid.model.EditSubscriptionModel
+import at.bitfire.icsdroid.model.SubscriptionSettingsModel
+import at.bitfire.icsdroid.ui.partials.ExtendedTopAppBar
+import at.bitfire.icsdroid.ui.partials.GenericAlertDialog
+import at.bitfire.icsdroid.ui.theme.AppTheme
+import at.bitfire.icsdroid.ui.views.LoginCredentialsComposable
+import at.bitfire.icsdroid.ui.views.SubscriptionSettingsComposable
+
+@Composable
+fun EditCalendarScreen(
+    application: Application,
+    subscriptionId: Long,
+    onShare: (subscription: Subscription) -> Unit,
+    onExit: () -> Unit = {}
+) {
+    val credentialsModel: CredentialsModel = viewModel()
+    val subscriptionSettingsModel: SubscriptionSettingsModel = viewModel()
+    val editSubscriptionModel: EditSubscriptionModel = viewModel {
+        EditSubscriptionModel(application, subscriptionId)
+    }
+    val editCalendarModel: EditCalendarModel = viewModel {
+        EditCalendarModel(editSubscriptionModel, subscriptionSettingsModel, credentialsModel)
+    }
+    EditCalendarScreen(
+        inputValid = editCalendarModel.inputValid,
+        modelsDirty = editCalendarModel.modelsDirty,
+        successMessage = editCalendarModel.editSubscriptionModel.uiState.successMessage,
+        onDelete = editSubscriptionModel::removeSubscription,
+        onSave = {
+            editSubscriptionModel.updateSubscription(subscriptionSettingsModel, credentialsModel)
+        },
+        {
+            editSubscriptionModel.subscriptionWithCredential.value?.let {
+                onShare(it.subscription)
+            }
+        },
+        onExit,
+        editCalendarModel.subscriptionSettingsModel.uiState.supportsAuthentication,
+        editCalendarModel.subscriptionSettingsModel,
+        editCalendarModel.credentialsModel
+    )
+}
+@Composable
+fun EditCalendarScreen(
+    inputValid: Boolean,
+    modelsDirty: Boolean,
+    successMessage: String?,
+    onDelete: () -> Unit,
+    onSave: () -> Unit,
+    onShare: () -> Unit,
+    onExit: () -> Unit,
+    supportsAuthentication: Boolean,
+    subscriptionSettingsModel: SubscriptionSettingsModel,
+    credentialsModel: CredentialsModel
+) {
```

Comment:
Since we want to pass down `ViewModel` instances as little as possible, in the other composables I've created a new function (in this case still named `EditCalendarScreen`), which receives all the arguments without any references to models, only raw types and callbacks.

As far as I understand it, maybe I'm wrong, but the idea is to have a Composable which receives the models, and then distributes the required data down.

It would mean calling the overload of `LoginCredentialsComposable` which takes arguments instead of the `CredentialsModel`, and the same goes for `SubscriptionSettingsComposable`.

However, this method causes functions to have a LOT of arguments, see `AddCalendarScreen`:

https://github.com/bitfireAT/icsx5/blob/1c618554949b43b6173ce8dcf82c45497d721b51/app/src/main/java/at/bitfire/icsdroid/ui/screen/AddCalendarScreen.kt#L179-L217

So even though I don't know if it's the best option, I believe it's the intended way to do it.

Reply:
Yeah, I saw that. And I was wondering whether to point it out or not. I don't think it makes the code more readable, which should be our top priority and is the reason for the guideline in the first case.

> Since we want to pass down ViewModel instances as little as possible

I think there's a sweet spot between too much and too little and we should always strive for clarity and readability than pleasing the guideline to the bottom. In the end it's just that - a guide, not a rule. 

In this case I think this **is** passing ViewModel as little as possible - without making the code blow up and become unreadable. 

I wonder if Ricki has something to say on this. May be worth asking. :)

Reply:
> I wonder if Ricki has something to say on this. May be worth asking. :)

Yeah, I think that since we have both cases now, we ask her and see what she thinks. She has also researched more than us on the topic, so will know more for sure.

Reply:
Didn't have a look in depth, but there's another reason for not passing models: Composables that rely on a Model can't be `@Preview`-ed anymore (because it's possible that the model can't be generated for previewing, and usually this is the case as far as I know). And usually we want Composables to just be visual representations of their state (which is passed in). Also the Composable can then do whatever evil things with the model and becomes a player that may do unexpected things.

So I'd say in general it's better to make things explicit and accept a few extra arguments rather than passing a model. (See https://developer.android.com/develop/ui/compose/state-hoisting#screen-ui-state)

Also there's usually (only) one model per screen, with [reusable things can be mixed in on domain layer over Use Cases](https://developer.android.com/topic/architecture/domain-layer). This may be helpful for better architecture. However there may of course be cases where multiple models are active at the same time. With NavGraph it's then also possible to scope a model above the screens to share data between them. I just want to note that using more than one model per screen seems to be a thing that needs to be considered twice :)

Reply:
I agree that maybe some models can be merged together, that would simplify a lot of things. Might be a thing for another PR though

Reply:
@ArnyminerZ Can you review again?